### PR TITLE
feat(ctb + challenger): Initial bond scaffolding

### PR DESCRIPTION
**Pull request by clabby**
## Overview

Adds the initial bond scaffolding in preparation for the bonding milestone. The goal of this PR isn't to implement bonding, but to:
* Prepare the `FaultDisputeGame` and `DisputeGameFactory` to accept ETH for bonded interactions.
* Pay out bonds to the proper recipient on claim resolution.
* Update the data structures to account for the bonds attached to bonded interactions.
* Add initial logic to the challenger to consult the `FaultDisputeGame` contract to fetch the bond amount for a given interaction kind and send this value along with bonded interactions.

**Metadata**
closes https://github.com/ethereum-optimism/client-pod/issues/422

### CODE REVIEW

**Comment by clabby** [comment]
Current dependencies on/for this PR:
* `develop`
  * **PR #8810** <a href="https://app.graphite.dev/github/pr/ethereum-optimism/optimism/8810?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
    * **PR #8811** <a href="https://app.graphite.dev/github/pr/ethereum-optimism/optimism/8811?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
      * **PR #8816** <a href="https://app.graphite.dev/github/pr/ethereum-optimism/optimism/8816?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈
        * **PR #8821** <a href="https://app.graphite.dev/github/pr/ethereum-optimism/optimism/8821?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
          * **PR #8822** <a href="https://app.graphite.dev/github/pr/ethereum-optimism/optimism/8822?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
            * **PR #8831** <a href="https://app.graphite.dev/github/pr/ethereum-optimism/optimism/8831?utm_source=stack-comment-icon" target="_blank">Graphite</a> 

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/ethereum-optimism/optimism/8816?utm_source=stack-comment).

File: packages/contracts-bedrock/src/dispute/FaultDisputeGame.sol [hunk 1]
```diff
@@ -534,6 +562,20 @@ contract FaultDisputeGame is IFaultDisputeGame, Clone, ISemver {
...
+    /// @return success_ Whether or not the transfer was successful.
+    function _payBond(address _recipient, ClaimData storage _bonded) internal returns (bool success_) {
+        // Set all bits in the bond value to indicate that the bond has been paid out.
+        uint256 bond = _bonded.bond;
+        if (bond == type(uint128).max) revert ClaimAlreadyResolved();
+        _bonded.bond = type(uint128).max;
+
+        // Pay out the bond to the recipient.
+        (success_,) = address(_recipient).call{ value: bond }(hex"");
+    }
```
Comment (refcell) [inline packages/contracts-bedrock/src/dispute/FaultDisputeGame.sol]: I'm much more in favor of using the pull vs push pattern here. This will undoubtedly be a security concern in any given audit.

File: packages/contracts-bedrock/test/actors/FaultDisputeActors.sol [hunk 2]
```diff
@@ -389,10 +405,14 @@ contract HonestDisputeActor is DisputeActor {
...
+            (bool innerSuccess,) = address(GAME).call{ value: localMove.value }(localMove.data);
             assembly {
                 success_ := and(success_, innerSuccess)
             }
         }
     }
+
+    fallback() external payable { }
+
+    receive() external payable { }
```
Comment (refcell) [inline packages/contracts-bedrock/test/actors/FaultDisputeActors.sol:417]: Since both are a no-op `receive` zero-calldata calls will be caught by `fallback` anyways, so only need to define the `fallback` selector here.
```suggestion
    fallback() external payable { }
```

File: packages/contracts-bedrock/test/dispute/FaultDisputeGame.t.sol [hunk 3]
```diff
@@ -80,19 +80,29 @@ contract FaultDisputeGame_Init is DisputeGameFactory_Init {
         // Label the proxy
         vm.label(address(gameProxy), "FaultDisputeGame_Clone");
     }
+
+    fallback() external payable { }
+
+    receive() external payable { }
```
Comment (refcell) [inline packages/contracts-bedrock/test/dispute/FaultDisputeGame.t.sol:86]: ibid.
```suggestion
    fallback() external payable { }
```

File: packages/contracts-bedrock/test/invariants/FaultDisputeGame.t.sol [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+        VM.deal(address(this), _bondAmount);
+
+        totalBonded += _bondAmount;
+
+        GAME.move{ value: _bondAmount }(_parentIndex, _claim, _isAttack);
+    }
+
+    fallback() external payable { }
+
+    receive() external payable { }
```
Comment (refcell) [inline packages/contracts-bedrock/test/invariants/FaultDisputeGame.t.sol:88]: ibid.
```suggestion
    fallback() external payable { }
```

File: packages/contracts-bedrock/test/actors/FaultDisputeActors.sol (hunk 2 above)
Comment (clabby) [reply packages/contracts-bedrock/test/actors/FaultDisputeActors.sol:417]: This produces a warning, solc 🫠

File: packages/contracts-bedrock/src/dispute/FaultDisputeGame.sol (hunk 1 above)
Comment (clabby) [reply packages/contracts-bedrock/src/dispute/FaultDisputeGame.sol]: Definitely something to discuss - I think that the push pattern is pretty safe here, and the solvency invariant is pretty clear to test. Ticketed out for further discussion https://github.com/ethereum-optimism/client-pod/issues/429

File: op-challenger/game/fault/responder/responder.go [hunk 5]
```diff
@@ -30,14 +32,19 @@ type FaultResponder struct {
...
+func NewFaultResponder(logger log.Logger, txMgr txmgr.TxManager, contract GameContract, gameDepth, splitDepth uint64) (*FaultResponder, error) {
 	return &FaultResponder{
-		log:      logger,
-		txMgr:    txMgr,
-		contract: contract,
+		log:        logger,
+		txMgr:      txMgr,
+		contract:   contract,
+		gameDepth:  gameDepth,
+		splitDepth: splitDepth,
```
Comment (ajsutton) [inline op-challenger/game/fault/responder/responder.go]: Why did we need the depths here? I don't see them being used.

File: op-challenger/game/fault/responder/responder.go [hunk 6]
```diff
@@ -92,6 +99,13 @@ func (r *FaultResponder) PerformAction(ctx context.Context, action types.Action)
 		} else {
 			candidate, err = r.contract.DefendTx(uint64(action.ParentIdx), action.Value)
 		}
+
+		// We can use a generic attack here, the bond calculation only focuses on the depth of the move position
+		bondValue, err := r.contract.GetRequiredBond(ctx, action.ParentPosition.Attack())
```
Comment (ajsutton) [inline op-challenger/game/fault/responder/responder.go]: Given we have an if statement just above to switch on attacking or not it's probably better to use the appropriate attack/defend case rather than always using attack. Otherwise we're making an assumption about the contract behaviour that isn't really part of its API contract.

Comment (ajsutton) [reply op-challenger/game/fault/responder/responder.go]: Or maybe even just add a method to `types.Action` that returns the position to use?

File: op-challenger/game/fault/responder/responder.go [hunk 7]
```diff
@@ -92,6 +99,13 @@ func (r *FaultResponder) PerformAction(ctx context.Context, action types.Action)
...
+		if err != nil {
+			return err
+		}
+		candidate.Value = bondValue
 	case types.ActionTypeStep:
 		candidate, err = r.contract.StepTx(uint64(action.ParentIdx), action.IsAttack, action.PreState, action.ProofData)
```
Comment (ajsutton) [inline op-challenger/game/fault/responder/responder.go:107]: Just to confirm, we don't need to pay a bond when calling step because it's the last step and we're incurring all the costs and not creating future costs for others.

File: op-challenger/game/fault/solver/game_solver.go [hunk 8]
```diff
@@ -50,7 +52,7 @@ func (s *GameSolver) CalculateNextActions(ctx context.Context, game types.Game)
 }
 
 func (s *GameSolver) calculateStep(ctx context.Context, game types.Game, agreeWithRootClaim bool, claim types.Claim) (*types.Action, error) {
-	if claim.Countered {
+	if claim.CounteredBy != common.BigToAddress(big.NewInt(0)) {
```
Comment (ajsutton) [inline op-challenger/game/fault/solver/game_solver.go]: Is this just `common.Address{}` or does it come out with something different?

File: op-e2e/e2eutils/disputegame/game_helper.go [hunk 9]
```diff
@@ -158,7 +158,7 @@ func (g *FaultGameHelper) WaitForClaimAtMaxDepth(ctx context.Context, countered
 		fmt.Sprintf("Could not find claim depth %v with countered=%v", maxDepth, countered),
 		func(claim ContractClaim) bool {
 			pos := types.NewPositionFromGIndex(claim.Position)
-			return int64(pos.Depth()) == maxDepth && claim.Countered == countered
+			return int64(pos.Depth()) == maxDepth && (claim.CounteredBy != common.BigToAddress(big.NewInt(0))) == countered
```
Comment (ajsutton) [inline op-e2e/e2eutils/disputegame/game_helper.go]: as above can this just be `common.Address{}`? If not we probably should create a constant for it since this is a fairly weird invocation.

File: op-e2e/e2eutils/disputegame/game_helper.go [hunk 10]
```diff
@@ -167,7 +167,7 @@ func (g *FaultGameHelper) WaitForAllClaimsCountered(ctx context.Context) {
 		ctx,
 		"Did not find all claims countered",
 		func(claim ContractClaim) bool {
-			return !claim.Countered
+			return claim.CounteredBy == common.BigToAddress(big.NewInt(0))
```
Comment (ajsutton) [inline op-e2e/e2eutils/disputegame/game_helper.go]: As above... I'm going to stop commenting on these but there's a couple more below. :)

File: packages/contracts-bedrock/invariant-docs/FaultDisputeGame.md [hunk 11]
```diff
@@ -0,0 +1,6 @@
+# `FaultDisputeGame` Invariants
+
+## FaultDisputeGame always returns all ETH on total resolution
+**Test:** [`FaultDisputeGame.t.sol#L38`](../test/invariants/FaultDisputeGame.t.sol#L38)
+
+The FaultDisputeGame contract should always return all ETH in the contract to the correct recipients upon resolution of all outstanding claims. There may never be any ETH left in the contract after a full resolution.
```
Comment (ajsutton) [inline packages/contracts-bedrock/invariant-docs/FaultDisputeGame.md:6]: Can we create a ticket to setup monitoring for this? Should be easy to check it in challenger and log a warning if a resolved game has funds remaining or something like that.

File: packages/contracts-bedrock/src/dispute/FaultDisputeGame.sol (hunk 1 above)
Comment (ajsutton) [reply packages/contracts-bedrock/src/dispute/FaultDisputeGame.sol]: Is the key concern here about the potential for payments to allow re-entrancy? Is there a risk that the payment fails and causes the whole tx to revert?

File: packages/contracts-bedrock/src/dispute/FaultDisputeGame.sol [hunk 12]
```diff
@@ -184,7 +184,7 @@ contract FaultDisputeGame is IFaultDisputeGame, Clone, ISemver {
 
         // Set the parent claim as countered. We do not need to append a new claim to the game;
         // instead, we can just set the existing parent as countered.
-        parent.countered = true;
+        parent.counteredBy = msg.sender;
```
Comment (ajsutton) [inline packages/contracts-bedrock/src/dispute/FaultDisputeGame.sol:190]: Should we allow people to set the address they want bonds paid to that are different from the source? Don't want to scope creep for a nice to have, but thinking of things like smart contract wallets or someone who's using a contract to batch multiple calls to the game.

**Review by ajsutton — COMMENTED** [review]
Haven't reviewed the contract changes carefully but the go code looks good generally with just a couple of small things to clean up.

File: op-challenger/game/fault/responder/responder.go (hunk 5 above)
Comment (clabby) [reply op-challenger/game/fault/responder/responder.go]: Ah good catch - artifact of before I swapped over to passing `Position` into the contract function. Removed.

File: op-challenger/game/fault/solver/game_solver.go (hunk 8 above)
Comment (clabby) [reply op-challenger/game/fault/solver/game_solver.go]:
```go
if claim.CounteredBy != common.Address{} {
    // ...
}
```
This actually doesn't compile :(

File: op-e2e/e2eutils/disputegame/game_helper.go (hunk 9 above)
Comment (clabby) [reply op-e2e/e2eutils/disputegame/game_helper.go]: This one can, other doesn't compile.

File: packages/contracts-bedrock/src/dispute/FaultDisputeGame.sol (hunk 12 above)
Comment (clabby) [reply packages/contracts-bedrock/src/dispute/FaultDisputeGame.sol:190]: Can ticket this out; Imagine that most people will be able to wire up logic to extract the bond payment from the contract that submits them worst-case, don't think it's important to add now.

File: packages/contracts-bedrock/invariant-docs/FaultDisputeGame.md (hunk 11 above)
Comment (clabby) [reply packages/contracts-bedrock/invariant-docs/FaultDisputeGame.md:6]: https://github.com/ethereum-optimism/client-pod/issues/433

File: packages/contracts-bedrock/src/dispute/FaultDisputeGame.sol (hunk 12 above)
Comment (clabby) [reply packages/contracts-bedrock/src/dispute/FaultDisputeGame.sol:190]: https://github.com/ethereum-optimism/client-pod/issues/434

File: op-challenger/game/fault/solver/game_solver.go (hunk 8 above)
Comment (ajsutton) [reply op-challenger/game/fault/solver/game_solver.go]: ugh, sorry, go is a bit daft...
```suggestion
	if claim.CounteredBy != (common.Address{}) {
```

**Review by ajsutton — APPROVED** [review]

File: packages/contracts-bedrock/src/dispute/FaultDisputeGame.sol (hunk 1 above)
Comment (clabby) [reply packages/contracts-bedrock/src/dispute/FaultDisputeGame.sol]: Swapped to pull

**Comment by clabby** [comment]
Assigning review DRIs to @ajsutton for Go code and @Inphi for the contracts. Adrian has context over the modifications to the `op-challenger`, and Mofi worked on the initial bonds specification.

File: packages/contracts-bedrock/invariant-docs/FaultDisputeGame.md (hunk 11 above)
Comment (Inphi) [reply packages/contracts-bedrock/invariant-docs/FaultDisputeGame.md:6]: should clarify that in the case someone sends ETH to the FDG, that leftover ETH will be burned.

File: packages/contracts-bedrock/src/dispute/FaultDisputeGame.sol [hunk 13]
```diff
@@ -391,15 +404,19 @@ contract FaultDisputeGame is IFaultDisputeGame, Clone, ISemver {
...
             // Ignore false claims
-            if (!claim.countered) {
-                countered = true;
+            if (claim.counteredBy == address(0)) {
+                countered = msg.sender;
                 break;
             }
         }
 
+        // If the parent was not successfully countered, pay out the parent's bond to the claimant.
```
Comment (Inphi) [inline packages/contracts-bedrock/src/dispute/FaultDisputeGame.sol:413]: this is good. I just want to explicitly call out that in the case there's a counter, we indeed want to reward the parent's bond to the _leftmost_ claimant, and not just any.

**Review by Inphi — APPROVED** [review]
Nice!

File: op-challenger/game/fault/contracts/faultdisputegame_test.go [hunk 14]
```diff
@@ -109,19 +109,23 @@ func TestGetClaim(t *testing.T) {
...
 	status, err := game.GetClaim(context.Background(), idx.Uint64())
 	require.NoError(t, err)
 	require.Equal(t, faultTypes.Claim{
 		ClaimData: faultTypes.ClaimData{
 			Value:    value,
 			Position: faultTypes.NewPositionFromGIndex(position),
+			Bond:     bond,
 		},
-		Countered:           true,
+		CounteredBy:         common.Address{0x01},
```
Comment (refcell) [inline op-challenger/game/fault/contracts/faultdisputegame_test.go]: Should use the predefined variable here
```suggestion
		CounteredBy:         counteredBy,
```

File: op-challenger/game/fault/responder/responder_test.go [hunk 15]
```diff
@@ -289,3 +290,7 @@ func (m *mockContract) UpdateOracleTx(_ context.Context, claimIdx uint64, data *
 	m.updateOracleArgs = data
 	return txmgr.TxCandidate{TxData: ([]byte)("updateOracle")}, nil
 }
+
+func (m *mockContract) GetRequiredBond(_ context.Context, position types.Position) (*big.Int, error) {
+	return big.NewInt(5), nil
+}
```
Comment (refcell) [inline op-challenger/game/fault/responder/responder_test.go:296]: We should add a test here that an error returned by fetching the required bond is bubbled up by `PerformAction`. Can do in a follow-up pr if you agree

**Review by refcell — APPROVED** [review]
This is awesome scaffolding, let's get this in and keep iterating 😄

File: op-challenger/game/fault/responder/responder_test.go (hunk 15 above)
Comment (clabby) [reply op-challenger/game/fault/responder/responder_test.go:296]: Yeah can follow up on this!

**Review by refcell — APPROVED** [review]
Subsequent changes look good 👍

File: packages/contracts-bedrock/test/actors/FaultDisputeActors.sol (hunk 2 above)
[the remark this answers is not in the document]
Comment (refcell) [reply packages/contracts-bedrock/test/actors/FaultDisputeActors.sol:417]: solc bugs out with a warning if removed

File: op-challenger/game/fault/responder/responder_test.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (refcell) [reply op-challenger/game/fault/responder/responder_test.go:296]: I frontran you
